### PR TITLE
[GlobalOpt] Recursively MaterializeHomogeneousEncoding on modules

### PR DESCRIPTION
**Pull request by Max191**
If there are nested modules, SetEncoding will act on all modules, but MaterializeHomogeneousEncoding will only act on the top level module. This PR allows MaterializeHomogeneousEncoding to recursively act on all nested modules.

### CODE REVIEW

**Review by qedawkins — CHANGES_REQUESTED** [review]
This seems like the wrong solution. Why are we ever adding `linalg_ext.encoding` to operations within a dispatch to begin with?

File: compiler/src/iree/compiler/GlobalOptimization/test/materialize_homogeneous_encodings.mlir [hunk 1]
```diff
@@ -96,3 +96,70 @@ module attributes {hal.device.targets = [#device_target_vulkan, #device_target_l
 // CHECK-LABEL: func.func @lhs_encoding
 // CHECK-SAME:    %[[ARG0:[a-zA-Z0-9]+]]
 // CHECK:         return %[[ARG0]]
+
+// -----
+
+#executable_target_rocm_hsaco_fb = #hal.executable.target<"rocm", "rocm-hsaco-fb", {target_arch = "gfx90a", ukernels = "none"}>
```
Comment (benvanik) [inline compiler/src/iree/compiler/GlobalOptimization/test/materialize_homogeneous_encodings.mlir]: this is a lot of IR for something that doesn't need it all - can you reduce this?
(every additional line of IR we have in lit tests that aren't relevant to the test becomes a burden if anything changes)

File: compiler/src/iree/compiler/GlobalOptimization/MaterializeHomogeneousEncodings.cpp [hunk 2]
```diff
@@ -76,6 +76,23 @@ class MaterializeHomogeneousEncodingsPass
       return signalPassFailure();
     }
   }
+
+  void runOnOperation() override {
+    auto rootOp = getOperation();
+    SetVector<ModuleOp> modules;
+    modules.insert(rootOp);
+    Explorer e(rootOp, TraversalAction::RECURSE);
+    e.walk([&modules](Operation *op) -> WalkResult {
```
Comment (benvanik) [inline compiler/src/iree/compiler/GlobalOptimization/MaterializeHomogeneousEncodings.cpp]: you can just do rootOp->walk to find the nested modules, right?

**Comment by Max191** [comment]
> This seems like the wrong solution. Why are we ever adding `linalg_ext.encoding` to operations within a dispatch to begin with?

SetEncoding happens before dispatch region formation. This bug occurs for any nested module, not just this testcase that happens to be a dispatch, so we need to make sure MaterializeEncoding happens on all modules in the program. I should make a better testcase, though.

**Comment by benvanik** [comment]
I think quinn's point is that set_encoding shouldn't be happening within preformed dispatch regions - what's the "bug"?

**Comment by Max191** [comment]
> I think quinn's point is that set_encoding shouldn't be happening within preformed dispatch regions - what's the "bug"?

If there is any module inside of another module, then we need to materialize encodings on it. There is a separate issue that we are setting encodings inside things like hal.executable, but we still need to make sure we recursively search modules here. I will file an issue about the issue with SetEncoding for later. I also updated the test to make it more clear what this is trying to fix.

**Comment by benvanik** [comment]
The only cases we should have nested modules is when they are flow/stream/hal executables, and if we don't want to have set_encoding in there, what's the case where we want it?

**Comment by benvanik** [comment]
(I'm not trying to be difficult, I'd just like to understand what you're doing - it shouldn't be needed, and if there is a legit use case there's _a lot_ of other things that need to be fixed)

**Comment by qedawkins** [comment]
We just chatted offline and the bug is in `SetEncoding`; we shouldn't ever be setting encodings on operations already inside a dispatch. @benvanik we were looking for a parent op trait to check for, and the one I thought of is `Util::ObjectLike`. Is that the right one to use?

**Comment by Max191** [comment]
> The only cases we should have nested modules is when they are flow/stream/hal executables, and if we don't want to have set_encoding in there, what's the case where we want it?

A user could try to compile something that has 2 modules inside of it, in which case there will be an outer module containing both of the modules inside of it.

**Comment by benvanik** [comment]
> A user could try to compile something that has 2 modules inside of it, in which case there will be an outer module containing both of the modules inside of it.

Yeah, there's _a lot_ needed to make that work, and I don't think a hack in this one pass is a good idea. IREE expects one compilation unit at a time and multiple nested modules like that is not valid. If anything, we should have a pass that runs as part of input processing and ensures no nested modules.

**Comment by Max191** [comment]
> Yeah, there's _a lot_ needed to make that work, and I don't think a hack in this one pass is a good idea. IREE expects one compilation unit at a time and multiple nested modules like that is not valid. If anything, we should have a pass that runs as part of input processing and ensures no nested modules.

Ok, makes sense. I'll close this and instead open up an issue about the issue with SetEncoding, then.

**Comment by benvanik** [comment]
#16411 will make MaterializeHomogenousEncodings only run on util.initializer/util.func, and those don't show up in executables, so there should be no issue then.

**Pull request closed without merging**